### ISSUE TITLE
Anonymous logon exception when creating new data source, uploading file, etc

### ISSUE THREAD

**Issue by bdwakefield** [Microsoft/Reporting-Services#23]
Custom security extension implemented as instructed in readme and is "working". I am able to login and create folders with the admin account entered in the config file and registered into DB. When I attempt to create a new data source; the following error shows up in the log file:

```
ERROR: Throwing Microsoft.ReportingServices.Diagnostics.Utilities.AuthenticationExtensionException: , Microsoft.ReportingServices.Diagnostics.Utilities.AuthenticationExtensionException: The Authentication Extension threw an unexpected exception or returned a value that is not valid: . ---> System.NullReferenceException: Anonymous logon is not configured. userIdentity should not be null!
   at Microsoft.Samples.ReportingServices.CustomSecurity.AuthenticationExtension.GetUserInfo(IIdentity& userIdentity, IntPtr& userId)
   at Microsoft.ReportingServices.WebServer.WebServiceHelper.ConstructRSServiceObjectFromSecurityExtension()
   --- End of inner exception stack trace ---;
```

For reference; exception is thrown from this code:

```
public void GetUserInfo(out IIdentity userIdentity, out IntPtr userId)
       {
           // If the current user identity is not null,
           // set the userIdentity parameter to that of the current user 
           if (HttpContext.Current != null
                 && HttpContext.Current.User != null)
           {
               userIdentity = HttpContext.Current.User.Identity;
           }
           else
           // The current user identity is null. This happens when the user attempts an anonymous logon.
           // Although it is ok to return userIdentity as a null reference, it is best to throw an appropriate
           // exception for debugging purposes.
           // To configure for anonymous logon, return a Gener
           {
               //System.Diagnostics.Debug.Assert(false, "Warning: userIdentity is null! Modify your code if you wish to support anonymous logon.");
               throw new NullReferenceException("Anonymous logon is not configured. userIdentity should not be null!");
           }
           // initialize a pointer to the current user id to zero
           userId = IntPtr.Zero;
       }
```

My understanding is that we have already logged in and authenticated and thus should NOT have an anonymous logon situation. I am not sure I want to modify the code to "handle" anonymous access -- as everyone should be authenticated; especially those creating/modifying report objects. If it matters; this is to provide external reporting through a custom web application.

I have replicated the same issue in a development VM running Windows 10; but the server details are below.

- Windows Server 2016 Standard
- SQL Server Standard v13.0.1601.5, Default Instance. Running under Network Service account.
- SSRS configured for HTTPS

**Comment by figuerres** [comment]
@bdwakefield   my take on this subject is this:   the sample code from Microsoft is just a starting point and fails to address the full scope of the problem.  I tried to use it and saw that issues like yours are going to be there and i stopped and want with a different method to run reports.

if you look at the total code in SSRS for user auth and how SSRS works it has a lot of things that depend on a user mapping to an AD user and SSRS roles mapping to an NTFS ACL object.
they do not really address the ACL's and the way that SSRS creates and presents them to the rest of the api's

**Comment by jtarquino** [comment]
Double check that your web.config has 
<authentication mode="Forms">
	<forms loginUrl="logon.aspx" name="sqlAuthCookie" timeout="60" path="/"></forms>
</authentication>

**Comment by bdwakefield** [comment]
I finally got it all working; I think the documentation needs improved. The thing that was missing was the cookie passthrough entries. The way the documentation reads; it is optional "when additional things are needed".

In my mind; being able to create more than just folders is not something that is "additional".

Another issue that I ran into that I do not have a good fix for (isn't related to this which is a documentation problem IMO) is the forms redirection losing query string parameters.

**Comment by bdwakefield** [comment]
@figuerres I actually have this working now; and I think it can be a viable option for embeding reports for external consumption. It was a lot more hoops than I had originally expected.

I am hoping to replicate everything I did and document it all. That was the biggest struggle I had was finding folks who had done it and succeed. All I could find was places where they got stuck and as things typically go they didn't ever get back to reporting what they did to fix it (or if they even gave up).

Our use case is embedding the report and providing access to the parameters of the report for additional filtering and exporting of the data. I didn't want to rebuild the report interface if I didn't have to. SSRS does a good job of taking care of that part of it.

**Comment by jtarquino** [comment]
@bdwakefield feel free to send a pull request with any update in the documentation those are more than welcome

**Comment by bdwakefield** [comment]
Will do; I want to walk through what I did again start to finish without the fumbling to make sure it is repeatable.

**Comment by npmuller** [comment]
@bdwakefield @figuerres I would appreciate if either of you could share your solution or solution architecture with me, I am having authentication issues when integrating into an iframe as well.  Thank you!

**Comment by npmuller** [comment]
For anyone reading this thread in the future: my solution was extraordinarily stupid and hopefully it will save you the 1.5 days I wasted on this:
- Followed the readme.md setup guide [here](https://github.com/microsoft/Reporting-Services/tree/master/CustomSecuritySample) to a T.
- Created a user using the C# code in this gist [here](https://gist.github.com/npmuller/c0484d17f1a8d6323afa874018635e10) and enter the username and hashed password/SALT generated by the into the UserAccounts.Users table created in step 1 of the readme above.

This is where I got stuck, for some reason the sqlAuthCookie was not being set on my client and I thought it may have had something to do with the fact that I was making a cross domain request (like bdwakefield and figuerres I inserting my report server into a custom web app by pointing the `src` attribute of an iframe at the report server).  Turns out the solution was stupidly simple:

**Don't test your webapp in a private browsing window!**

That's why the sqlAuthCookie wasn't getting set, because NO cookies were getting set.  I was using FF 69.0.3.  When I tested the site in a non-private browser window, everything worked fine.

**Comment by bdwakefield** [comment]
> @bdwakefield @figuerres I would appreciate if either of you could share your solution or solution architecture with me, I am having authentication issues when integrating into an iframe as well. Thank you!

I ended up not repeating what we did. It was ugly and I was not confident in the answer. I will say I never had any issues with private browsing. I am curious about it preventing the cookie from being created -- it seems like private browsing should do everything as the normal one would with the only difference being when you close the browser everything is deleted.

Our "hack" was on the page in our application that lists the available reports (we have a table that lists the reports and SSRS URLs, clicking the report link in our application opens a page that puts that path into the iFrame); we make a quick web request to the auto-login page (with hardcoded 'viewer' credentials). Since our primary issue users not getting authed. It is not great and I am not happy with it... but it has been working for a few years with zero issues.
